File: Medium/valid_sudoku/valid_sudoku.py
from typing import List

def isValidSudoku(self, board: List[List[str]]) -> bool:
    for i in range(9):
        for j in range(9):
            num = board[i][j]
            if num == '.':
                continue
            # Check row
            for k in range(9):
                if k != j and board[i][k] == num:
                    return False
            # Check column
            for k in range(9):
                if k != i and board[k][j] == num:
                    return False
            start_row = (i // 3) * 3
            start_col = (j // 3) * 3
            for r in range(start_row, start_row + 3):
                for c in range(start_col, start_col + 3):
                    if (r != i or c != j) and board[r][c] == num:
                        return False
    return True

if __name__ == "__main__":
    print("Testing isValidSudoku function:")
    board1 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    print("For the first board, the output is:", isValidSudoku(None, board1)) #Output: True

    board2 = [
        ["8","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    print("For the second board, the output is:", isValidSudoku(None, board2)) #Output: False
        
        

File: Medium/valid_sudoku/valid_sudoku.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
    public:
        bool isValidSudoku(vector<vector<char>>& board){
            for(int i = 0; i < 9; i++){
                unordered_set<char> seen;
                for(int j = 0; j < 9; j++){
                    char current = board[i][j];
                    if(current == '.'){
                        continue;
                    }
                    if(seen.count(current)){
                        return false;
                    }
                    seen.insert(current);
                }
            }
            for(int j = 0; j < 9; j++){
                unordered_set<char> seen;
                for(int i = 0; i < 9; i++){
                    char current = board[i][j];
                    if(current == '.'){
                        continue;
                    }
                    if(seen.count(current)){
                        return false;
                    }
                    seen.insert(current);
                }
            }
            for(int boxCol = 0; boxCol < 9; boxCol = boxCol + 3) {
                for(int boxRow = 0; boxRow < 9; boxRow = boxRow + 3){
                    unordered_set<char> seen;
                    for(int i = 0; i < 3; i++){
                        for(int j = 0; j < 3; j++){
                            char current = board[boxRow + i][boxCol + j];
                            if(current == '.'){
                                continue;
                            }
                            if(seen.count(current)){
                                return false;
                            }
                            seen.insert(current);
                        }
                    }
                }
            }
        return true;
    }
};

// Example usage:
int main(){
    Solution solution;
    vector<vector<char>> board = {
        {'5','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };
    cout << boolalpha; 
    bool result = solution.isValidSudoku(board);
    cout << "For the first board, the Sudoku board output is " << result << "." << endl;

    vector<vector<char>> board2 = {
        {'8','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };
    cout << boolalpha;
    bool result2 = solution.isValidSudoku(board2);
    cout << "For the second board, the Sudoku board output is " << result2 << "." << endl;
}